File: iosApp/iosApp/Canvas/SurfaceCanvas.swift
import SwiftUI
import SharedLogic

/// 2D elevation canvas drawing helpers for surface tile layouts.
struct SurfaceCanvas {
    /// Draw all tiles for a surface onto the given Canvas context.
    static func drawTiles(
        context: inout GraphicsContext,
        size: CGSize,
        tiles: [PlacedTile],
        groutColor: Color,
        groutWidth: Double
    ) {
        guard !tiles.isEmpty else { return }

        // Compute the bounding box of all tiles to determine scale
        let maxX = tiles.map { $0.x + $0.width }.max() ?? 1
        let maxY = tiles.map { $0.y + $0.height }.max() ?? 1
        let scaleX = size.width / maxX
        let scaleY = size.height / maxY
        let scale = min(scaleX, scaleY) * 0.95
        let offsetX = (size.width - maxX * scale) / 2
        let offsetY = (size.height - maxY * scale) / 2

        // Draw grout background
        let bgRect = CGRect(x: offsetX, y: offsetY, width: maxX * scale, height: maxY * scale)
        context.fill(Path(bgRect), with: .color(groutColor))

        // Draw each tile
        for tile in tiles {
            let tileRect = CGRect(
                x: offsetX + tile.x * scale,
                y: offsetY + tile.y * scale,
                width: tile.width * scale,
                height: tile.height * scale
            )

            // Fill tile with texture or fallback color
            let tileColor = tileFillColor(tile: tile)
            context.fill(Path(tileRect), with: .color(tileColor))

            // Draw tile border
            context.stroke(Path(tileRect), with: .color(.black.opacity(0.3)), lineWidth: 0.5)

            // Mark cut edges
            if tile.isCut {
                drawCutMarkers(context: &context, rect: tileRect, cutEdges: tile.cutEdges as? [CutEdge] ?? [])
            }

            // Label rotation for herringbone
            if abs(tile.rotation) > 0.01 {
                let center = CGPoint(x: tileRect.midX, y: tileRect.midY)
                let text = Text("\(Int(tile.rotation))°")
                    .font(.system(size: 8))
                    .foregroundColor(.white)
                context.draw(text, at: center)
            }
        }
    }

    private static func tileFillColor(tile: PlacedTile) -> Color {
        if tile.isCut {
            return Color.orange.opacity(0.5)
        }
        // Color by tileGroupId hash for visual distinction
        let hash = tile.tileGroupId.hashValue
        let hue = Double(abs(hash) % 360) / 360.0
        return Color(hue: hue, saturation: 0.3, brightness: 0.85)
    }

    private static func drawCutMarkers(
        context: inout GraphicsContext,
        rect: CGRect,
        cutEdges: [CutEdge]
    ) {
        let markerWidth: CGFloat = 3.0
        for edge in cutEdges {
            let markerRect: CGRect
            switch edge {
            case CutEdge.left:
                markerRect = CGRect(x: rect.minX, y: rect.minY, width: markerWidth, height: rect.height)
            case CutEdge.right:
                markerRect = CGRect(x: rect.maxX - markerWidth, y: rect.minY, width: markerWidth, height: rect.height)
            case CutEdge.top:
                markerRect = CGRect(x: rect.minX, y: rect.minY, width: rect.width, height: markerWidth)
            case CutEdge.bottom:
                markerRect = CGRect(x: rect.minX, y: rect.maxY - markerWidth, width: rect.width, height: markerWidth)
            default:
                continue
            }
            context.fill(Path(markerRect), with: .color(.red))
        }
    }
}

File: androidApp/src/main/kotlin/com/hasu/tilelayout/ui/canvas/SurfaceCanvas.kt
package com.hasu.tilelayout.ui.canvas

import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.geometry.Size
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.PathEffect
import androidx.compose.ui.graphics.drawscope.DrawScope
import androidx.compose.ui.graphics.drawscope.Stroke
import com.hasu.tilelayout.models.CutEdge
import com.hasu.tilelayout.models.GroutColor
import com.hasu.tilelayout.models.PlacedTile
import com.hasu.tilelayout.models.RegionRect

// Grout color mapping
fun GroutColor.toComposeColor(): Color = when (this) {
    GroutColor.BLACK -> Color.Black
    GroutColor.GREY -> Color.Gray
    GroutColor.WHITE -> Color.White
}

// Draw all tiles on a surface
fun DrawScope.drawTiles(
    tiles: List<PlacedTile>,
    groutColor: GroutColor = GroutColor.GREY,
    groutWidth: Double = 3.0,
    scale: Float = 1f,
    doorRect: RegionRect? = null,
) {
    val grout = groutColor.toComposeColor()
    for (tile in tiles) {
        val x = (tile.x * scale).toFloat()
        val y = (tile.y * scale).toFloat()
        val w = (tile.width * scale).toFloat()
        val h = (tile.height * scale).toFloat()

        // Draw grout (slightly larger rect)
        val g = groutWidth.toFloat() * scale
        drawRect(color = grout, topLeft = Offset(x - g, y - g), size = Size(w + 2 * g, h + 2 * g))

        // Draw tile body
        val tileColor = if (tile.isCut) Color(0xFFE8D5B7) else Color(0xFFD4A574) // lighter for cuts
        drawRect(color = tileColor, topLeft = Offset(x, y), size = Size(w, h))

        // Draw cut edge indicators
        for (edge in tile.cutEdges) {
            val lineColor = Color.Red
            when (edge) {
                CutEdge.LEFT -> drawLine(lineColor, Offset(x, y), Offset(x, y + h), strokeWidth = 1.5f)
                CutEdge.RIGHT -> drawLine(lineColor, Offset(x + w, y), Offset(x + w, y + h), strokeWidth = 1.5f)
                CutEdge.TOP -> drawLine(lineColor, Offset(x, y), Offset(x + w, y), strokeWidth = 1.5f)
                CutEdge.BOTTOM -> drawLine(lineColor, Offset(x, y + h), Offset(x + w, y + h), strokeWidth = 1.5f)
            }
        }
    }

    // Dashed door outline drawn after the tiles so the door area reads as explicit
    if (doorRect != null) {
        drawRect(
            color = Color.Gray,
            topLeft = Offset(
                (doorRect.x * scale).toFloat(),
                (doorRect.y * scale).toFloat(),
            ),
            size = Size(
                (doorRect.width * scale).toFloat(),
                (doorRect.height * scale).toFloat(),
            ),
            style = Stroke(
                width = 1.5f,
                pathEffect = PathEffect.dashPathEffect(floatArrayOf(8f, 6f)),
            ),
        )
    }
}
